programs/clearing_house/src/lib.rs

```diff
@@ -754,6 +757,226 @@ pub mod clearing_house {
+
+        let update_k_result = get_update_k_result(&market, new_sqrt_k_u192, false)?;
+        math::amm::update_k(&mut market, &update_k_result)?;
+
+        Ok(())
+    }
+
+    #[access_control(
+        exchange_not_paused(&ctx.accounts.state)
+    )]
+    pub fn add_liquidity<'info>(
+        ctx: Context<AddRemoveLiquidity>,
+        quote_asset_amount: u128,
+        market_index: u64,
+    ) -> Result<()> {
+        let user = &mut load_mut(&ctx.accounts.user)?;
+        let clock = Clock::get()?;
+        let remaining_accounts_iter = &mut ctx.remaining_accounts.iter().peekable();
+
+        let mut oracle_map = OracleMap::load(remaining_accounts_iter, clock.slot)?;
+        let bank_map = BankMap::load(
+            &get_writable_banks(QUOTE_ASSET_BANK_INDEX),
+            remaining_accounts_iter,
+        )?;
+        let market_map = MarketMap::load(
+            &get_writable_markets(market_index),
+            &MarketOracles::new(), // TODO
+            remaining_accounts_iter,
+        )?;
+
+        let position_index = get_position_index(&user.positions, market_index)
+            .or_else(|_| add_new_position(&mut user.positions, market_index))?;
+        let lp_position = &mut user.positions[position_index];
+
+        // TODO: allow lps to add to position when settle is working
+        let can_lp = !(lp_position.has_open_order()
+            || lp_position.is_open_position()
+            || lp_position.is_lp());
+        validate!(can_lp, ErrorCode::CantLPWithMarketPosition)?;
+
+        let (
+            peg_multiplier,
+            net_base_asset_amount,
+            total_fee_minus_distributions,
+            cumulative_funding_rate_lp,
+            sqrt_k,
+        ) = get_struct_values!(
+            market_map.get_ref(&market_index)?.amm,
+            peg_multiplier,
+            net_base_asset_amount,
+            total_fee_minus_distributions,
+            cumulative_funding_rate_lp,
+            sqrt_k
+        );
+
+        // create lp position
+        // TODO: change from peg to mark price?
+        let user_lp_tokens = quote_asset_amount
+            .checked_mul(AMM_TO_QUOTE_PRECISION_RATIO)
+            .ok_or_else(math_error!())?
+            .checked_div(PEG_PRECISION)
+            .ok_or_else(math_error!())?
+            .checked_div(2)
+            .ok_or_else(math_error!())?
+            .checked_div(peg_multiplier)
+            .ok_or_else(math_error!())?;
+
+        // init position
+        lp_position.last_net_base_asset_amount = net_base_asset_amount;
+        lp_position.last_total_fee_minus_distributions = total_fee_minus_distributions;
+        lp_position.last_cumulative_funding_rate = cumulative_funding_rate_lp;
+
+        // add token balance
+        lp_position.lp_tokens = lp_position
+            .lp_tokens
+            .checked_add(user_lp_tokens)
+            .ok_or_else(math_error!())?;
+
+        // update market state
+        let new_sqrt_k = sqrt_k
```

Comment:
should we have guardrails around how much a single user can lp? or around how much of total liquidity can be from lps?